stages/org.osbuild.coreos.live-artifacts.mono

```diff
@@ -0,0 +1,845 @@
+        with tempfile.TemporaryDirectory() as d:
+            try:
+                subprocess.check_call(['mount', '-o', 'utf8', loopdev, d])
+                subprocess.check_call(['tar', '-C', d, '-xf', input_tarball])
+            finally:
+                subprocess.check_call(['umount', d])
+
+
+def parse_metal_inputs(inputs):
+    def get_filepath_from_input(name):
+        files = inputs[name]["data"]["files"]
+        assert len(files) == 1
+        filename, _ = files.popitem()
+        filepath = os.path.join(inputs[name]["path"], filename)
+        return filepath
+    metal_file = get_filepath_from_input('metal')
+    metal4k_file = get_filepath_from_input('metal4k')
+    return metal_file, metal4k_file
+
+
+# Here we will generate the 4 Live/PXE CoreOS artifacts.
+# There are four files created and exported from this stage.
+#
+# 1. live-iso
+# 2. live-kernel    (i.e. images/pxeboot/vmlinuz)
+# 3. live-initramfs (i.e. images/pxeboot/initrd.img)
+# 4. live-rootfs    (i.e. images/pxeboot/rootfs.img)
+#
+# The live-iso itself contains the other 3 artifacts inside of
+# it. A rough approximation of the structure looks like:
+#
+#   live-iso.iso
+#     -> images/pxeboot/vmlinuz
+#     -> images/pxeboot/initrd.img
+#     -> images/pxeboot/rootfs.img
+#       -> metal.osmet
+#       -> metal4k.osmet
+#       -> root.squashfs
+#         -> full filesystem of metal image
+#
+# Where the ISO contains the kernel and initrd and the rootfs. The
+# rootfs contains the osmet file and the root.squashfs, which itself
+# contains all the files from a full CoreOS system.
+# pylint: disable=too-many-statements,too-many-branches
+def main(workdir, tree, inputs, options, loop_client):
+    squashfs_compression = 'zstd'
+    basearch = os.uname().machine
+    filenames = options['filenames']
+    output_iso = os.path.join(tree, filenames['live-iso'])
+    output_kernel = os.path.join(tree, filenames['live-kernel'])
+    output_rootfs = os.path.join(tree, filenames['live-rootfs'])
+    output_initramfs = os.path.join(tree, filenames['live-initramfs'])
+    img_metal, img_metal4k = parse_metal_inputs(inputs)
+    # The deployed tree input is a deployment tree just as it would
+    # appear on a booted OSTree system. We copy some files out of this
+    # input tree since it is easier and they match what is in the metal
+    # images anyway. We also use it as a chroot when we run `coreos-installer`.
+    deployed_tree = inputs['deployed-tree']['path']
+
+    # For dev/test purposes it is useful to create test fixture ISO
+    # images for the coreos-installer CI framework to use [1]. This test
+    # fixture is much smaller than an actual ISO and can be stored in git.
+    # Determine if the user wants a test fixture ISO created by checking
+    # for the coreos-installer-test-fixture file baked in the tree.
+    # [1] https://github.com/coreos/coreos-installer/tree/main/fixtures/iso
+    test_fixture = os.path.exists(
```

Comment:
> Can we somehow leverage this for our tests here?

I don't think so?? In OSBuild CI IIUC you don't really do much with the generated artifacts. i.e. you don't boot the images you generate and run tests on them. You just make sure the artifacts can be generated.

Is my understanding correct?

This test fixture is more specifically to test [coreos-installer](https://github.com/coreos/coreos-installer) in the future for regressions. 

> 
> Also, having this unconditionally in our tree makes me a bit nervous as it may lead to building with test artificats unintentionally, can we somehow encode this in the manifest without making your life harder?

I talked with @jlebon about this. I really didn't want to expose something in the schema (i.e. something that would need to be put into the manifest) that was specific to testing an upstream project. We discussed not putting this code in here at all, but co-locating it is the best way to ensure we are generating a test artifact that is very close to the real thing. 

The solution we came up with was to only generate these test fixture artifacts if the payload had a file at `usr/share/coreos-assembler/coreos-installer-test-fixture`. This enables a `coreos-installer` developer to easily create a CoreOS build with that file existing and thus a test artifact ISO, but no one else needs to know about it.

In my mind this is similar to someone setting an environment variable like `COREOS_LIVE_ARTIFACTS_TEST_FIXTURE_ISO=1` and then running OSBuild, but passing in an env variable like this into a stage clearly won't work (i.e. locked down `bwrap`) without some significant plumbing, which I don't really think is what we want either.

Reply:
Thanks for the detailed response. I agree there is no great way to set it otherwise, maybe we should invest into this, in the meantime it is fine (even though I still don't like it :)